src/server/store/ObjectStore.ts

```diff
@@ -0,0 +1,133 @@
+  byAttributed: AbstractLevel<any, string, any>
+  byReply: AbstractLevel<any, string, any>
+  byPublished: AbstractLevel<any, string, any>
+
+  constructor (db: AbstractLevel<any, string, any>) {
+    this.db = db
+    this.objects = db.sublevel('objects', { valueEncoding: 'json' })
+    this.byAttributed = db.sublevel('byAttributedTo')
+    this.byReply = db.sublevel('byInReplyTo')
+    this.byPublished = db.sublevel('byPublished')
+  }
+
+  urlToKey (url: string): string {
+    // URL encode the url to clean up the special chars before inserting
+    return encodeURIComponent(url)
+  }
+
+  async add (object: APObject): Promise<void> {
+    if (typeof object.id !== 'string') throw createError(400, 'Objects must have id fields')
+    const hasPublished = typeof object.published === 'string'
+    const hasInReplyTo = typeof object.inReplyTo === 'string'
+    const hasAttributedTo = typeof object.attributedTo === 'string'
+    if (hasPublished) {
+      await this.byPublished
+        .sublevel(object.published as string)
+        .put(object.id, object.id)
+    }
+    if (hasPublished && hasInReplyTo) {
+      await this.byReply
+        .sublevel(object.inReplyTo as string)
+        .sublevel(object.published as string)
+        .put(object.id, object.id)
+    }
+    if (hasPublished && hasAttributedTo) {
+      await this.byAttributed
+        .sublevel(object.attributedTo as string)
+        .sublevel(object.published as string)
+        .put(object.id, object.id)
+    }
+    await this.objects.put(this.urlToKey(object.id), object)
+  }
+
+  async remove (url: string): Promise<void> {
+    const object = await this.get(url)
+    const hasPublished = typeof object.published === 'string'
+    const hasInReplyTo = typeof object.inReplyTo === 'string'
+    const hasAttributedTo = typeof object.attributedTo === 'string'
+    if (hasPublished) {
+      await this.byPublished
+        .sublevel(object.published as string)
+        .del(url)
+    }
+    if (hasPublished && hasInReplyTo) {
+      await this.byReply
+        .sublevel(object.inReplyTo as string)
+        .sublevel(object.published as string)
+        .del(url)
+    }
+    if (hasPublished && hasAttributedTo) {
+      await this.byAttributed
+        .sublevel(object.attributedTo as string)
+        .sublevel(object.published as string)
+        .del(url)
+    }
+    await this.objects.del(this.urlToKey(url))
+  }
+
+  async get (url: string): Promise<APObject> {
+    const key = this.urlToKey(url)
+    try {
+      const note: APObject = await this.objects.get(key)
+      return note
+    } catch (error) {
+      throw createError(404, `Object not found for URL: ${url}`)
+    }
+  }
+
+  async list ({ skip = 0, limit = Infinity, attributedTo, inReplyTo, to }: ListParameters = {}): Promise<APObject[]> {
+    const items: APObject[] = []
+
+    let iterator = null
+
+    if (inReplyTo !== undefined) {
+      iterator = this.byReply.sublevel(inReplyTo).values({ reverse: true })
```

Comment:
So is this the method for detecting the `inReplyTo` field in a post to link it to the original reply?

Reply:
I think so yes. https://docs.joinmastodon.org/spec/activitypub/#$:~:text=inReplyTo